README.md
plugin.program.hubhug.maze3
===========================

HUB-HUG Exit The Maze 3:  Sleeping Monsters
===========================
$  - User

S  - Start

E  - Exit / End

SPACE  - Path

\#  - Wall

L  - Life

K  - Key

D  - Door

R  - Monster0R

0  - Monster00

1  - Monster01

2  - Monster02

3  - Monster03

4  - Monster04

5  - Monster05

6  - Monster06

7  - Monster07

8  - Monster08

9  - Monster09


===========================

Monster Battle Wins/Losses are not currently setup.

When Life hits 0, you loose and you exit from the game automatically.
When you hit the EXIT / END, the game will automatically start another level.

You can only pass a DOOR when you have 1 or more key(s).  A key is used up every time you use one to pass through a door.  What can I say, they break EASILY!

Please note that there are Sound FX(s) in this game.  So adjust volume settings accordingly in XBMC and/or your OS.



Addon Settings
===========================

Player Avatar  - Select the graphic you'll play as.

Select Map  - There is an option in Addon Settings to let you manually browse/select a map.

Show map  - Show Text type Map.

/\

Show location on the map  - Update Text type map as you move.

Show items  - Display items and monsters even when not next to them.


===========================


changelog.txt
[B]Version 0.2.2[/B]
[I]06-11-2014[/I]
- Adjustments to File Updater.

[B]Version 0.2.1[/B]
[I]06-11-2014[/I]
- File Updater.

[B]Version 0.2.0[/B]

[I]06-11-2014[/I]
- Some sturdy testing for map issues and base optoins.
- Additional customizable values.
- Additional sounds.
- Text Map options auto calculated according to the permitted Width values for the Visual Display.
- Changes to base options.
- Changes to Settings.xml.

[B]Version 0.1.8[/B]
- SUPER MODE looses it's color for the last two moves before it turns off.
- SUPER MODE setup and ready.
- Fix for Display of Game Stats.
- Attempt @ Battle Coding.
- Additional Game Stats.
- SaveGame (stats) now has a yes/no popup to ask if you want to save.
- LoadGame (stats) on initial startup of game client.
- SaveGame (stats) on exit.
- Additoinal Customizable settings for map files.

[B]Version 0.1.6[/B]
- Fixed Left/Right movement of mobs.

[B]Version 0.1.5[/B]
- Facing Directions for SOME of the USER graphics.
- More Monster Graphics.
- More USER Graphics.

[B]Version 0.1.2[/B]

[I]06-08-2014[/I]
- Made a Fun little wide open map for testing movement coding and such.
- Some slight straight-directional Monster Movement within so much distance of the user.  Currently no diagnal movement.  Currently Monsters only see in a straight line.  Currently Monsters don't start a battle with you, unless you move onto their position.
- Text Map's View Distance.
- Extra Layer of Images for Floor Graphic underneath transparent game graphics.  This way other floor/path styles can be used at times.
- Customizable Graphic,Sound,Text Message settings in map files.  Some settings are customizable, but not all at this time.
- Added Additional Graphics.
- Initial Setup for Direction the User is facing, for later coding.
- Fixed some graphics.
- Changed some graphics.

[I]06-??-2014[/I]
- ....
- ....

[B]Version 0.1.0[/B]
- Many advancements.

[I]06-??-2014[/I]
- ....
- ....

[I]05-??-2014[/I]
- ....
- ....

[B]Version 0.0.1[/B]
- Added HUB HUG Movement Series Fanart Graphic by Marquerite.
- Added Some Graphic(s) by TheHighway.
- Initial Build.



common.py
## ################################################## ##
## ################################################## ##
## Common functions for use by TheHighway.            ##
## ################################################## ##
## ################################################## ##
## import common as Common
## from common import *
## import common
## ################################################## ##
## ################################################## ##

import xbmc,xbmcgui,urllib,urllib2,os,sys,logging,array,re,time,datetime,random,string,StringIO,xbmcplugin,xbmcaddon
#import shutil
from config import Config as Config
MyBrowser=['User-Agent','Mozilla/5.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.3) Gecko/2008092417 Firefox/3.0.3']; 

def cFLL( t,c='tan'): return '[ccccc_'+c+']'+t+'[/ccccc]' ### For Coloring Text ###
def cFL( t,c='tan'): return '[COLOR '+c+']'+t+'[/COLOR]' ### For Coloring Text ###
def cFL_(t,c='tan'): return '[COLOR '+c+']'+t[0:1]+'[/COLOR]'+t[1:] ### For Coloring Text (First Letter-Only) ###
def deb(s,t): ### for Writing Debug Data to log file ###
	#if (_debugging==True): 
	print s+':  '+t
def debob(t): ### for Writing Debug Object to log file ###
	#if (_debugging==True): 
	print t
def nolines(t):
	it=t.splitlines(); t=''
	for L in it: t=t+L
	t=((t.replace("\r","")).replace("\n",""))
	return t
def art(f,fe=''): 
	fe1='.png'; fe2='.jpg'; fe3='.gif'; 
	if   fe1 in f: f=f.replace(fe1,''); fe=fe1; 
	elif fe2 in f: f=f.replace(fe2,''); fe=fe2; 
	elif fe3 in f: f=f.replace(fe3,''); fe=fe3; 
	return xbmc.translatePath(os.path.join(Config.artPath,f+fe))
def artp(f,fe='.png'): 
	return art(f,fe)
def artj(f,fe='.jpg'): 
	return art(f,fe)
def addonPath(f,fe=''):
	return xbmc.translatePath(os.path.join(Config.path,f+fe))
def BusyAnimationShow(): 				xbmc.executebuiltin('ActivateWindow(busydialog)')
def BusyAnimationHide(): 				xbmc.executebuiltin('Dialog.Close(busydialog,true)')
def closeAllDialogs():   				xbmc.executebuiltin('Dialog.Close(all, true)') 
def popYN(title='',line1='',line2='',line3='',n='',y=''):
	diag=xbmcgui.Dialog()
	r=diag.yesno(title,line1,line2,line3,n,y)
	if r: return r
	else: return False
	#del diag
def popOK(msg="",title="",line2="",line3=""):
	dialog=xbmcgui.Dialog()
	#ok=dialog.ok(title, msg, line2, line3)
	dialog.ok(title, msg, line2, line3)
def note(title='',msg='',delay=5000,image='http://upload.wikimedia.org/wikipedia/commons/thumb/a/a5/US_99_%281961%29.svg/40px-US_99_%281961%29.svg.png'): xbmc.executebuiltin('XBMC.Notification("%s","%s",%d,"%s")' % (title,msg,delay,image))
def repo(repository,filename='',h="special://home",a="addons"): 
	if len(filename)==0: return xbmc.validatePath(xbmc.translatePath(os.path.join(h,a,repository)))
	else: return xbmc.validatePath(xbmc.translatePath(os.path.join(h,a,repository,filename)))
def repox(repository,filename='',h="special://xbmc",a="addons"): 
	if len(filename)==0: return xbmc.validatePath(xbmc.translatePath(os.path.join(h,a,repository)))
	else: return xbmc.validatePath(xbmc.translatePath(os.path.join(h,a,repository,filename)))
def _SaveFile(path,data):
	file=open(path,'w')
	file.write(data)
	file.close()
def _OpenFile(path):
	#deb('File',path)
	if os.path.isfile(path): ## File found.
		#deb('Found',path)
		file = open(path, 'r')
		contents=file.read()
		file.close()
		return contents
	else: return '' ## File not found.
def _CreateDirectory(dir_path):
	dir_path = dir_path.strip()
	if not os.path.exists(dir_path): os.makedirs(dir_path)
def _get_dir(mypath, dirname): #...creates sub-directories if they are not found.
	subpath = os.path.join(mypath, dirname)
	if not os.path.exists(subpath): os.makedirs(subpath)
	return subpath
def getURL(url):
	try:
		req=urllib2.Request(url)
		req.add_header(MyBrowser[0],MyBrowser[1]) 
		response=urllib2.urlopen(req)
		link=response.read()
		response.close()
		return(link)
	except: deb('Failed to fetch url',url); return ''
def postURL(url,form_data={},headers={},compression=True):
	try:
		req=urllib2.Request(url)
		if form_data: form_data=urllib.urlencode(form_data); req=urllib2.Request(url,form_data)
		req.add_header(MyBrowser[0],MyBrowser[1])
		for k, v in headers.items(): req.add_header(k, v)
		if compression: req.add_header('Accept-Encoding', 'gzip')
		response=urllib2.urlopen(req)
		link=response.read()
		response.close()
		return link
	except: deb('Failed to fetch url',url); return ''
def postURL2(url,form_data={}):
	try:
		postData=urllib.urlencode(form_data)
		req=urllib2.Request(url,postData)
		req.add_header(MyBrowser[0], MyBrowser[1]) 
		response=urllib2.urlopen(req)
		link=response.read()
		response.close()
		return(link)
	except: deb('Failed to fetch url',url); return ''
def setContent(content,h=0):
	h=Config.handle
	try: 		xbmcplugin.setContent(h,content)
	except: pass
def SettingG(setting):
	try: return Config.addon.getSetting(setting)
	except: return ""
def SettingS(setting,value): Config.addon.setSetting(id=setting,value=value)
def tfalse(r,d=False): ## Get True / False
	if   (r.lower()=='true' ) or (r.lower()=='t') or (r.lower()=='y') or (r.lower()=='1') or (r.lower()=='yes'): return True
	elif (r.lower()=='false') or (r.lower()=='f') or (r.lower()=='n') or (r.lower()=='0') or (r.lower()=='no'): return False
	else: return d
def showkeyboard(txtMessage="",txtHeader="",passwordField=False):
	try:
		if txtMessage=='None': txtMessage=''
		keyboard = xbmc.Keyboard(txtMessage, txtHeader, passwordField)#("text to show","header text", True="password field"/False="show text")
		keyboard.doModal()
		if keyboard.isConfirmed(): return keyboard.getText()
		else: return '' #return False
	except: return ''
def is_odd(num): 
	try:
		return num % 2 != 0
	except: return False
def DoA(a): xbmc.executebuiltin("Action(%s)" % a)
def SFX(n,e='.wav'):
	if len(n)==0: return
	if (n==' ') or (n=='...') or (n=='_'): return
	snd=art(n,e)
	try: xbmc.playSFX(snd,False)
	except: 
		try: xbmc.playSFX(snd)
		except: pass
def isPath(path): return os.path.exists(path)
def isFile(filename): return os.path.isfile(filename)
## ################################################## ##
def CatchLocalFileList():
	zz=[xbmc.translatePath(Config.path),xbmc.translatePath(Config.resourcesPath),xbmc.translatePath(Config.puzzlePath),xbmc.translatePath(Config.artPath)]
	zL=''
	for z in zz:
		try:
			zList=os.listdir(z)
			for L in zList:
				if (not '.bak' in L) and (not '.pyo' in L) and (not '.log' in L) and (not '.lnk' in L) and (not 'thumbnail' in L.lower()):
					if '.' in L:
						try: zL+=os.path.join(z,L)+'\n'; 
						except: pass
		except: pass
	zL=zL.replace(Config.path+'\\','').replace(Config.path,'').replace('\\','|').replace('\n\n','\n').strip()
	_SaveFile(addonPath('temp.txt'),zL)
def ExtractThis(filename,destpath):
	import Extract as extract
	return extract.allNoProgress(filename,destpath)
def UpdateCheck():
	if tfalse(SettingG("updater"))==False: return
	P=xbmc.translatePath(Config.path); U=Config.UpdateListUrl; 
	print U+Config.UpdateListFile; 
	#try:
	List_Local =_OpenFile(addonPath(Config.UpdateListFile)).strip().replace('\a','\n').replace('\r','\n').replace('\n\n\n','\n').replace('\n\n','\n').strip(); 
	print str(len(List_Local)); 
	List_Online=getURL(U+Config.UpdateListFile).strip().replace('\a','\n').replace('\r','\n').replace('\n\n\n','\n').replace('\n\n','\n').strip(); 
	print str(len(List_Online)); 
	if List_Local==List_Online: print 'Files are up to date.'; return
	if not '||' in List_Online: print 'Error checking for updates.  Error in online file list.'; return
	#except: print 'Error checking for updates.'; return
	import HiddenDownloader as MyDownloader
	DP=xbmcgui.DialogProgress(); DP.create('Updating Files','Initializing...'); 
	try:
		#if '\n||\n' in List_Online: List_Online=List_Online.split('\n||\n')[0].strip()
		if '||' in List_Online: List_Online=List_Online.split('||')[0].strip()
	except: pass
	print str(len(List_Online)); 
	List_Array=List_Online.strip().split('\n')
	print str(len(List_Array)); print List_Array; 
	total=len(List_Array); curI=0; FilesFailed=0; 
	for FileItem in List_Array:
		FileItem=FileItem.strip(); 
		if (len(FileItem) > 1) and (not '#' in FileItem) and ('.' in FileItem):
			try: tFolder,tFile=FileItem.split('|'); 
			except: tFolder=''; tFile=FileItem; 
			DP.update((100*curI/total),'Path: '+tFolder,'File: '+tFile)
			time.sleep(2); 
			#print tFolder+'\\'+tFile; 
			if len(tFolder)==0: tLocalFile=xbmc.translatePath(os.path.join(P,tFile))
			else: tLocalFile=xbmc.translatePath(os.path.join(P,tFolder,tFile))
			if len(tFolder)==0: tLocalPath=xbmc.translatePath(P)
			else: tLocalPath=xbmc.translatePath(os.path.join(P,tFolder))
			if len(tFolder)==0: tRemoteFile=U+tFile
			else: tRemoteFile=U+tFolder+'/'+tFile
			print [str((100*curI/total)),str((curI/total*100)),tFolder,tFile,tLocalPath,tLocalFile,tRemoteFile]; 
			try:
				MyDownloader.download(tRemoteFile,tFile,tLocalPath,False); 
				if '.zip' in tFile: ExtractThis(tLocalFile,tLocalPath)
			#except urllib2.URLError, e: print e.code; 
			except: print 'error downloading file: '+tFile; FilesFailed+=1; 
		curI+=1; 
		print str(FilesFailed)+' files failed to download.'
	DP.close()
## ################################################## ##


## ################################################## ##
## ################################################## ##


splash_highway.py
### ####################################################################### ### 
### import splash_highway as splash
### splash.do_My_Splash('http://xbmchub.com/images/index-logo.png',5); 
### splash.do_My_Splash('http://xbmchub.com/images/index-logo.png',2,True,100,100,600,400); 
### ####################################################################### ### 
import xbmc,xbmcgui,time
### ####################################################################### ### 
class MyWindowCountDown(xbmcgui.WindowDialog):
	scr={}; #scr['L']=0; scr['T']=0; scr['W']=1280; scr['H']=720; 
	def __init__(self,bgArt='',L=0,T=0,W=1280,H=720):
		self.background=bgArt; self.scr['L']=L; self.scr['T']=T; self.scr['W']=W; self.scr['H']=H; 
		self.BG=xbmcgui.ControlImage(self.scr['L'],self.scr['T'],self.scr['W'],self.scr['H'],self.background,aspectRatio=0); 
		self.addControl(self.BG); 
		#t='0'; self.BG.setAnimations([('WindowOpen','effect=fade time='+t+' start=0 end=100'),('WindowClose','effect=fade time='+t+' end=0')])
		#self.close()
	def see(self): self.show(); 
	def updateBG(self,bgArt=''): self.background=bgArt; self.setImage(self.background); 
	def updateY(self,x,y): self.BG.setPosition(x,y); self.scr['L']=x; self.scr['T']=y; 
	def updateSize(self,w,h): self.BG.setWidth(w); self.BG.setHeight(h); self.scr['W']=w; self.scr['H']=h; 
	def updateW(self,w): self.BG.setWidth(w); self.scr['W']=w; 
	def updateH(self,h): self.BG.setHeight(h); self.scr['H']=h; 
	#def onInit(self): 
	#def onClick(self,control): 
	#def onControl(self,control): 
	#def onFocus(self,control): 
	#def onAction(self,action): 
def do_My_Splash(img='http://xbmchub.com/images/index-logo.png',HowLong=10,resize=False,L=0,T=0,W=1280,H=720): #HowLong in seconds.
	if resize==False: maxW=1280; maxH=720; W=maxW/2; H=maxH/2; L=maxW/4; T=maxH/4; 
	TempWindow2=MyWindowCountDown(bgArt=img,L=L,T=T,W=W,H=H); 
	StartTime=time.clock(); 
	while (time.clock()-StartTime) < HowLong:
		TempWindow2.show(); 
	try: del self.TempWindow2; 
	except: pass
### ####################################################################### ### 
class MyWindowCountDownWithText(xbmcgui.WindowDialog):
	scr={}; #scr['L']=0; scr['T']=0; scr['W']=1280; scr['H']=720; 
	def __init__(self,msg='',bgArt='',L=0,T=0,W=1280,H=720,TxtColor='0xFFFFFFFF',Font='font14',BorderWidth=10):
		self.background=bgArt; self.scr['L']=L; self.scr['T']=T; self.scr['W']=W; self.scr['H']=H; 
		self.BG=xbmcgui.ControlImage(self.scr['L'],self.scr['T'],self.scr['W'],self.scr['H'],self.background,aspectRatio=0); 
		self.addControl(self.BG); 
		self.TxtMessage=xbmcgui.ControlTextBox(self.scr['L']+BorderWidth,self.scr['T']+BorderWidth,self.scr['W']-(BorderWidth*2),self.scr['H']-(BorderWidth*2),font=Font,textColor=TxtColor); 
		self.addControl(self.TxtMessage); 
		self.TxtMessage.setText(msg); 
	def see(self): self.show(); 
	def updateBG(self,bgArt=''): self.background=bgArt; self.setImage(self.background); 
	def updateY(self,x,y): self.BG.setPosition(x,y); self.scr['L']=x; self.scr['T']=y; 
	def updateSize(self,w,h): self.BG.setWidth(w); self.BG.setHeight(h); self.scr['W']=w; self.scr['H']=h; 
	def updateW(self,w): self.BG.setWidth(w); self.scr['W']=w; 
	def updateH(self,h): self.BG.setHeight(h); self.scr['H']=h; 
	#def onInit(self): 
	#def onClick(self,control): 
	#def onControl(self,control): 
	#def onFocus(self,control): 
	#def onAction(self,action): 

def do_My_TextSplash(msg='',img='http://xbmchub.com/images/index-logo.png',HowLong=10,resize=False,L=0,T=0,W=1280,H=720,TxtColor='0xFFFFFFFF',Font='font14',BorderWidth=10): #HowLong in seconds.
	if resize==False: maxW=1280; maxH=720; W=maxW/2; H=maxH/2; L=maxW/4; T=maxH/4; 
	TempWindow2=MyWindowCountDownWithText(msg=msg,bgArt=img,L=L,T=T,W=W,H=H,TxtColor=TxtColor,Font=Font,BorderWidth=BorderWidth); 
	StartTime=time.clock(); 
	while (time.clock()-StartTime) < HowLong:
		TempWindow2.show(); 
	try: del self.TempWindow2; 
	except: pass
### ####################################################################### ### 


config.py
## from config import Config as Config

import xbmc,xbmcaddon,xbmcplugin,os
_addon_id="plugin.program.hubhug.maze3"; _addon_name="HUB-HUG Dungeon Crawler"; 
_addon=xbmcaddon.Addon(id=_addon_id); 

def gAI(t):
	try: return _addon.getAddonInfo(t)
	except: 
		#if t=="name": return _addon_name
		#else: return ""
		return ""

class Config:
	addon_id=_addon_id
	#name='Repositories Browser'
	addon=_addon
	handle=0
	name=gAI('name'); 
	name2=gAI('name') +":  "+"[COLOR seagreen]Rescuing My Love[/COLOR]"; 
	#name="XBMCHUB Repositories Browser"; 
	path=gAI('path'); 
	#profile=xbmc.translatePath(gAI('profile'))
	disclaimer=gAI('disclaimer'); description=gAI('description'); summary=gAI('summary'); 
	icon=gAI('icon'); fanart=gAI('fanart'); 
	addon_type=gAI('type'); author=gAI('author'); version=gAI('version'); stars=gAI('stars'); changelog=gAI('changelog'); 
	
	artPath=xbmc.translatePath(os.path.join(path,'art2'))
	
	
	ACTION_PREVIOUS_MENU 		=  10	## ESC action
	ACTION_NAV_BACK 				=  92	## Backspace action
	ACTION_MOVE_LEFT 				=   1	## Left arrow key
	ACTION_MOVE_RIGHT 			=   2	## Right arrow key
	ACTION_MOVE_UP 					=   3	## Up arrow key
	ACTION_MOVE_DOWN 				=   4	## Down arrow key
	ACTION_MOUSE_WHEEL_UP 	= 104	## Mouse wheel up
	ACTION_MOUSE_WHEEL_DOWN = 105	## Mouse wheel down
	ACTION_MOUSE_DRAG 			= 106	## Mouse drag
	ACTION_MOUSE_MOVE 			= 107	## Mouse move
	#
	ACTION_SELECT_ITEM			=		7	## ?
	ACTION_PARENT_DIR				=		9	## ?
	ACTION_CONTEXT_MENU			=	117	## ?
	ACTION_NEXT_ITEM				=	 14	## ?
	ACTION_BACKSPACE				=	110	## ?
	
	
	
	def gSetting(self,setting):
		try: return self.addon.getSetting(setting)
		except: return ""
	def set_setting(self,setting,value):
		self.addon.setSetting(id=setting,value=value)
	def get_string(self, string_id):
		try: return self.addon.getLocalizedString(string_id)
		except: return ""
	def get_id(self):
		try: return self.addon.getAddonInfo('id')
		except: return self.addon_id
	def note(self,title='',msg='',delay=5000,image='http://upload.wikimedia.org/wikipedia/commons/thumb/a/a5/US_99_%281961%29.svg/40px-US_99_%281961%29.svg.png'):
		xbmc.executebuiltin('XBMC.Notification("%s","%s",%d,"%s")' % (title,msg,delay,image))
	def show_settings(self): self.addon.openSettings()
	def eod(self):
		xbmcplugin.endOfDirectory(self.handle)
	
	
	
	
	